### ISSUE TITLE
are there any quick wins for reducing the install size of nyc?

### ISSUE THREAD

**Issue by jescalan** [istanbuljs/nyc#459]
When installed, this package and its dependencies come out to more than 25mb, which is really quite large. Just wanted to open up a conversation to see if perhaps there's any way to start slimming it down!

**Comment by bcoe** [comment]
@jescalan I certainly wouldn't complain about having a second set of eyes. Having said this, we've done quite a bit of analysis:

1. a big chunk of the size is related to istanbul and babel, which we can't easily do away with since it's how everything works.
2. we bundle all our dependencies for nyc; even though the final size on disk is large, nyc comes down as a tarball in a single HTTP request, so the install time is quite fast.
3. 25MB isn't _that big_ for a development tool; would certainly stink if there was a compelling reason to ship nyc as a front-end package -- but I don't think this will ever be a use-case?

**Comment by bcoe** [comment]
@jescalan were you interested in digging into this a bit?

**Comment by jescalan** [comment]
Yeah so I poked around a little, and it seems like babel is a large part of it (with its dependency on core-js). I guess there is no way to remove this if you want to provide coverage for es6 code, right?

**Comment by bcoe** [comment]
@jescalan correct, and I definitely feel like switching to using babel as the instrumenter has been beneficial to the project... we could dig into https://github.com/istanbuljs/istanbul-lib-instrument, and see if any of the babel dependencies can be trimmed down, we could also dive into the babel project itself and see if there are any tricks we could use to reduce the code size.

**Comment by Flux159** [comment]
I was doing a quick check of my node modules folder and saw that nyc was one of the largest dependencies and saw that this issue had already been opened.

I was wondering if there had been any progress on this - I did a few quick checks to count the number of files and size of each directory and found that nyc downloaded 4000 files with a total size of 25mb.

```
> find ./node_modules/nyc/ -type f | wc -l
4000
```

Listing the top 20 node_modules gave the following result:
```
> du -sk ./node_modules/nyc/node_modules/* | sort -rn | head -n 20
6752	./node_modules/nyc/node_modules/core-js
4964	./node_modules/nyc/node_modules/lodash
3048	./node_modules/nyc/node_modules/handlebars
964	./node_modules/nyc/node_modules/babel-runtime
784	./node_modules/nyc/node_modules/uglify-js
784	./node_modules/nyc/node_modules/source-map
528	./node_modules/nyc/node_modules/yargs
292	./node_modules/nyc/node_modules/babylon
216	./node_modules/nyc/node_modules/babel-traverse
172	./node_modules/nyc/node_modules/babel-types
172	./node_modules/nyc/node_modules/async
156	./node_modules/nyc/node_modules/istanbul-reports
144	./node_modules/nyc/node_modules/optimist
132	./node_modules/nyc/node_modules/babel-generator
104	./node_modules/nyc/node_modules/cliui
100	./node_modules/nyc/node_modules/debug
92	./node_modules/nyc/node_modules/mkdirp
80	./node_modules/nyc/node_modules/istanbul-lib-report
76	./node_modules/nyc/node_modules/istanbul-lib-instrument
76	./node_modules/nyc/node_modules/convert-source-map
```

Where the numbers are in kb. Doing some checks using npm ls showed this for the top 3:

```
> cd node_modules/nyc
> npm ls core-js
└─┬ istanbul-lib-instrument@1.7.2
  └─┬ babel-generator@6.24.1
    └─┬ babel-runtime@6.23.0
      └── core-js@2.4.1 
> npm ls lodash
├─┬ istanbul-lib-instrument@1.7.2
│ ├─┬ babel-generator@6.24.1
│ │ └── lodash@4.17.4  deduped
│ ├─┬ babel-template@6.24.1
│ │ └── lodash@4.17.4  deduped
│ ├─┬ babel-traverse@6.24.1
│ │ └── lodash@4.17.4  deduped
│ └─┬ babel-types@6.24.1
│   └── lodash@4.17.4  deduped
└── lodash@4.17.4 
> npm ls handlebars
└─┬ istanbul-reports@1.1.1
  └── handlebars@4.0.10 
```

Looking at istanbul-lib-instrument's package.json (https://github.com/istanbuljs/istanbuljs/blob/master/packages/istanbul-lib-instrument/package.json) it seems like its using older versions of babel which may make a difference in size, but I'm not entirely sure if there are easy wins here. I just wanted to check if someone anyone was still looking at this or looking to optimize the package. Thanks.

**Comment by coreyfarrell** [comment]
Is 61160771 still needed?  I just checked `node_modules/` and `node_modules/nyc/node_modules/`.  69 modules were installed in both locations.  Some may be required due to different versions, but those 69 modules are 7.4 megs in the nyc bundled copy.

**Comment by ORESoftware** [comment]
install it globally, or better yet, put it in user $HOME

```bash
#!/usr/bin/env bash
(
  mkdir $HOME/.my-nyc
  cd $HOME/.my-nyc
  npm init -f
  npm install -S nyc
)
```

then when you do your thing include that location in your $PATH

```bash
#!/usr/bin/env bash

export PATH="$HOME/.my-nyc/node_modules/.bin:${PATH}"
# now run your tests
nyc ...   # nyc will be the one installed in `$HOME/.my-nyc/node_modules/.bin`
```

**Comment by coreyfarrell** [comment]
@ORESoftware `nyc` gets listed in `devDependencies` because it is used by `npm test`.  Installing it globally would not do anything, the package would still install it locally as a dev dependency and use the local copy instead of the global one.

**Comment by ORESoftware** [comment]
@coreyfarrell who/what is putting `nyc` in `devDependencies` though?

**Comment by realityking** [comment]
I just looked into the same thing. To be honest, I'm not a fan of bundled dependencies, especially with super common ones as it's the case here. The original reason to do so was to ship a patched version of `spawn-wrap`. Since that's resolved, maybe it's time to consider that approach?

**Comment by coreyfarrell** [comment]
One thing I've noticed is that nyc seems to require node.js 4.x+.  I created a fork, enabled travis-ci on it with the following changes:
* I cleared the devDependencies.
* Set the script for `npm test` to `./bin/nyc.js ./bin/nyc.js --help`.

This command fails on nodejs 0.10 and 0.12.  The issue is due to use of ES2015 features both within `nyc` itself and in dependencies (`find-up` at minimum).

Am I missing something or is it acceptable to explicitly require node.js >= 4.0 so we can update remaining dependencies?  If so what about https://github.com/istanbuljs/istanbuljs/?  0.10 and 0.12 are EOL long ago, the oldest supported version of node.js is 6.x.  Updating dependencies will reduce duplication, in my own private install I've reduced the `npm install --prod` size by 16% (from 30MB to 25MB).  I've verified `npm test` still passes with the updated dependencies.

I still believe that dependencies should not be bundled without a technical reason.  Even if such a reason exists for specific modules most/all others should be unbundled?

**Comment by ORESoftware** [comment]
@coreyfarrell  why are you stuck on 0.10 or 0.12 ? Those are very old versions now.

**Comment by coreyfarrell** [comment]
@ORESoftware I'm not, I run 8.x.  My hope is that nyc and istanbuljs can drop support for those versions so dependencies can be brought up to date.

**Comment by ORESoftware** [comment]
Oh I see, that is a good point

**Comment by realityking** [comment]
> Is [6116077](https://github.com/istanbuljs/nyc/commit/61160771dd23cc2a6efa22d8fe77c286027df161) still needed? I just checked `node_modules/` and `node_modules/nyc/node_modules/`. 69 modules were installed in both locations. Some may be required due to different versions, but those 69 modules are 7.4 megs in the nyc bundled copy.

@coreyfarrell Did you ever find out if the bundled dependencies are needed? It would be love to get things to deduplicate. 🙂

**Comment by coreyfarrell** [comment]
@realityking Some dependencies may use global/singleton state.  A problem has occurred in the past where NYC used (directly or indirectly) a component which uses a singleton pattern.  Then the code being tested within NYC tries using the same and misbehaves because NYC changed the initial singleton state.

The largest module currently bundled by `nyc` is handlebars at 3.3mb.  This is used to produce HTML reports.  I suspect a smaller solution could be done but that would involve essentially rewriting the HTML report generator.  I'd be open to such a solution but it would have to be proven that produces the same output.  Also I don't have time to work on it.

Worth noting `babel` is no longer bundled with NYC as we had an issue with bundled `istanbul-lib-instrument` failing to install in recent versions.  I'd be open to unbundling other modules but it would need to be proven that unbundling cannot result in bugs caused by a singleton pattern anywhere in the dependency chain.  I don't know of any automated tool to scan for this.

Bottom line this is not a priority, bugs which directly cause run-time problems are a higher priority and we don't have enough time/help to address all of them.  Sorry, avoiding regressions is more important than reducing disk space requirements.
